練習：タブっぽいものを作る #21

mamiline6 opened the pull request:
@geckotang 

一通り作成してみました。
よろしくお願いします！

mamiline6 committed:
gulpfile の整理

#21用のタスク追加
ファイルの生成元・先変更

mamiline6 committed:
#21用 HTML作成

mamiline6 committed:
タブ切り替え用JS

自習

mamiline6 committed:
.gitignore ルートにあるscssディレクトリのみ無視

mamiline6 committed:
Revert ".gitignore ルートにあるscssディレクトリのみ無視"

This reverts commit b4af39fc970f21e36ec97cc0cffb4a15b2fd26dc.

mamiline6 committed:
gitignore ルートにあるscssディレクトリのみ無視

mamiline6 committed:
タブ用scss

mamiline6 committed:
reset,base.scss パーシャルで分割

mamiline6 committed:
gitignore 更新

ルートscssディレクトリのみを無視

geckotang commented on gulpfile.js:
```diff
@@ -1,14 +1,18 @@
+var rename       = require('gulp-rename');
+var minify       = require('gulp-minify-css');
 var autoprefixer = require('gulp-autoprefixer');
+var plumber      = require('gulp-plumber');　// watch の解除をしない命令文
```
セミコロンの後ろに全角空白が入っている

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,63 @@
+      border-width: 0 1px 0 0;
+      border-style: solid;
+      border-color: $borderColor;
+      background: #ddd;
```
色だけなら `background-color` を使うほうが好み。（必要のないショートハンドを使わない）

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,63 @@
+    text-align: left;
+  }
+  &.is-hidden {
+  display: none; /* js */
```
インデント

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,63 @@
+
+.tabContents {
+  padding: 40px;
+  border-width: 0 1px 1px;
```
`border-width`は4箇所全部書いたほうがわかりやすいと思っている。

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,63 @@
+  border-left: 1px solid $borderColor;
+  font-size: 0;
+  &__tab {
+    display: inline-block;
```
このタブは折り返す必要ないので、`table-cell`とかのほうがいいかもしれない。
そうすれば余計なコードが削れるので。

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,63 @@
+    *display: inline;
+    *zoom: 1;
+    width: 25%;
+    &.is-select a {
```
`selected`かな？

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,63 @@
+    &.is-select a {
+      background-color: #fff;
+    }
+    & a {
```
``` scss
.tabs {
  &__tab {

  }
  &__link {

  }
}
```

にすると、詳細度が下がって良いと思う。

geckotang commented on practice_tab/src/scss/_common.scss:
```diff
@@ -0,0 +1,27 @@
+body {
+	position: relative;
+	font-size: 10px;
+	text-align: center;
```
ターゲットが一応IE8以上なので、いらない気がする（やるにしても、いちいち戻さなくちゃいけないので`#wrapper`とかで`text-align: left;`するのがいいかもしれない。

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,63 @@
+  border-width: 0 1px 1px;
+  border-style: solid;
+  border-color: $borderColor;
+  & h2 {
```
`.tabContens`の中身は、きっと何が入っても良い想定なので、特に指定しなくて良いかもしれない。
`.tabContens`に依存しない別のモジュールにしちゃってもいいかも。

geckotang commented on practice_tab/app/js/style.js:
```diff
@@ -0,0 +1,10 @@
+$(function() {
+	// tab
+	$(".tabBlock .tabs__tab a").click(function() {
+		var num = $(".tabBlock .tabs__tab a").index(this);
```
`$(".tabBlock .tabs__tab a")`が2回出てきてるので、クリックする前に変数に入れておくほうが良い。
以下も同じ記述（`$(".tabContents")`、`$(".tabBlock .tabs__tab")など）は変数化。

geckotang commented on practice_tab/src/scss/_common.scss:
```diff
@@ -0,0 +1,27 @@
+fieldset,img {border: 0; vertical-align: bottom;}
+select option {padding: 0 5px;}
+form label {cursor: pointer;}
+article,aside,dialog,figure,footer,header,menu,nav,section {display: block;}
```
ここまでのリセット部分は、別のscssファイルにして、_common.scssにimportするようにしたら、使い回しが楽になりそう。

geckotang commented on practice_tab/app/js/style.js:
```diff
@@ -0,0 +1,10 @@
+	// tab
+	$(".tabBlock .tabs__tab a").click(function() {
+		var num = $(".tabBlock .tabs__tab a").index(this);
+		$(".tabContents").addClass('is-hidden');
```
もし、.tabBlockの塊が、ページ内にもう一つあった時、これだとそっちも閉じてしまうのです...!

``` jade
.tabBlock
  ul.tabs
    li.tabs__tab.is-select
    li.tabs__tab
  .tabContents.is-select
  .tabContents.is-hidden
.tabBlock
  ul.tabs
    li.tabs__tab.is-select
    li.tabs__tab
  .tabContents.is-select
  .tabContents.is-hidden
```

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,63 @@
+
+/* tabBlock */
+.tabs {
+  border-top: 2px solid $borderColor;
```
border-widthとかを使ったほうがわかりやすくなりそう

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,63 @@
+    & a {
+      display: block;
+      padding: 20px 0;
+      font-size: 16px;
```
`inline-block`にした時の、改行による空白を消すのに`font-size: 0;`を使うのであれば、liに`font-size: 16px;`を指定するほうがよい。（liのa要素だけに限定する必要はないので）

table-cellとかflexboxが使えるとまたコードは変わってくるですが。

mamiline6 committed:
セミコロンの後ろの全角空白を半角に修正

mamiline6 committed:
インデント変更

半角2スペースをタブに変更
minify()のコメントアウト削除

mamiline6 committed:
コードの修正

余計なスペースを削除
ブロックにid（#tabTopBlock）追加
タブのクラス名（.is-selected）修正

mamiline6 committed:
ファイル名変更

_common.scssからreset部分を分割

mamiline6 committed:
_common.scss 調整

_reset.scss部分以外、tab部分の指定以外をまとめる

mamiline6 committed:
style.scss修正

_reset.scssをインポート
tab部分のみのcss
tabをtable-cellで指定
borderプロパティの記述の修正
backgroundプロパティの記述の修正
aタグを__linkに修正
.tabContents内の指定は_common.scssに移植
.is-hiddenのインデント修正

mamiline6 committed:
[WIP] 全体のコード修正

繰り返し記述している部分を変数にまとめる
- [] .tabBlockが複数合ってもそれぞれに影響しないようにする

mamiline6 commented:
@geckotang 

たくさんご指摘ありがとうございます。

jsは範囲外でしたが見ていただきありがとうございます。
style.js で複数tabBlockが存在してもいいようにの解決方法が
id指定じゃなくて、何か配列などを使うのか思いつかなかったです。
100個あったら100個書かないといけないのが、とても現実的じゃないコードです。

プロパティの指定方歩、クラス名の付け方、
ファイルの分割について参考にします。

mamiline6 committed:
セミコロン抜けてました…

geckotang commented on practice_tab/app/js/style.js:
```diff
@@ -0,0 +1,15 @@
+
+	$(tabLink).click(function() {
+		var num = tabLink.index(this);
+		console.log(tabCnt);
```
console.log残ってる

geckotang commented on practice_tab/src/scss/_common.scss:
```diff
@@ -0,0 +1,27 @@
+	line-height: 1.5;
+	color: #333;
+	font-family: 'Hiragino Kaku Gothic ProN', Meiryo, 'MS PGothic', Sans-serif;
+	-webkit-text-size-adjust: none;
```
noneよりは100%のほうがいいかもめ

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,53 @@
+@charset "UTF-8";
+
+@import "_reset.scss";
+@import "_common.scss";
```
アンダースコアも拡張子もいらないきがするけど、好みだからこのままでも

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,53 @@
+		&:hover {
+			text-decoration: none;
+			background-color: #eee;
+			transition: 0.2s;
```
transition使うときは、`all`であっても明示したほうが伝わりやすい。
transition: 0.2s all linear;

geckotang commented on practice_tab/src/scss/style.scss:
```diff
@@ -0,0 +1,53 @@
+		}
+	}
+	& .is-selected .tabs__link {
+		background-color: #fff;
```
選択前の色と選択後の色も、変数化しておくとよい

geckotang commented on gulpfile.js:
```diff
@@ -1,30 +1,59 @@
-  gulp.watch(sassSrc, ['sass']);
-});
\ No newline at end of file
+	gulp.watch(sassSrc, ['sass']);
```
タブの方のCSSを監視するwatchタスクを作ってもいいと思うょ（めんどうじゃなかった...?）

mamiline6 committed:
console.log削除

mamiline6 committed:
タスクwatchを追記

mamiline6 committed:
PC版で拡大できるようにtext-size-adjustを修正

mamiline6 committed:
インポート指定を省略

mamiline6 committed:
transitionプロパティの指定を省略せず明示

geckotang commented:
@mamiline6 http://codepen.io/geckotang/pen/yNRpyy?editors=001 こんなかんじのJS書いてみたのでよければ見てみてね！！

mamiline6 committed:
タブの選択前の色と選択後の色も、変数化

mamiline6 committed:
style.js コピペ

.eachの理解

mamiline6 committed:
例にならって、コードの変更

mamiline6 committed:
例にならって、コードの変更

変数名がコメントなくてもわかりやすい
クラス名がより完結になって関連性がわかりやすい

mamiline6 commented:
最後、ただのコピペになってしまいましたが
ありがとうございました！
